### PR TITLE
Database Schema

### PR DESCRIPTION
**Pull request by bensoccer94**
Let me know if there are any changes that need to be made to the code.

Also, Im trying to figure out how to add an automatic delete date value for a record that gets removed from a table. This is for the deleted date columns.

Also, im having trouble properly setting up the correct CHECK constraint for the email address column. If anyone knows the solution, let me know and help me understand it lol :)

[SQL CODE.txt](https://github.com/martyboggs/video-service/files/4806907/SQL.CODE.txt)

### CODE REVIEW

**Comment by martyboggs** [comment]
BEN'S FIRST PULL REQUEEEEEEEEEEEEST!  LET'S GOOOOOOOOOO!

**Comment by martyboggs** [comment]
Add Nguyen and me as reviewers. It's on the top right.

**Comment by bensoccer94** [comment]
WOOOOOOOOHHOOOOOOOO LOL


I see a check mark on each of ur names.. can u see the code?

**Review by martyboggs — APPROVED** [review]
Excellent addition.

File: SQL CODE.txt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+
+CREATE TABLE users
+(
+	userId INT PRIMARY KEY IDENTITY(1,1),
+	first_name VARCHAR(40) NOT NULL,
+	last_name VARCHAR(40) NULL,
+	username VARCHAR(25) UNIQUE NOT NULL,
+	email VARCHAR(30) UNIQUE NOT NULL CHECK (email like '%_@%.com'),
+	[password] VARCHAR(15) NOT NULL,
```
Comment (nsnguyen) [inline SQL CODE.txt:12]: why did you put in [password]? if you can't do password, can you have another column name like user_password?

File: SQL CODE.txt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+CREATE TABLE video
+(
+	videoId INT PRIMARY KEY IDENTITY(1,1),
+	userId INT REFERENCES users(userId) NOT NULL,
+	Video_name VARCHAR(30) NOT NULL,
+	Video_description VARCHAR(MAX) NULL,
+	latitude float NULL,
+	longitude float NULL,
+	video_path VARCHAR(MAX) NOT NULL,
+	Video_Likes INT NOT NULL,
```
Comment (nsnguyen) [inline SQL CODE.txt:34]: video likes can be null i think right?

File: SQL CODE.txt [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	Video_dislikes INT NOT NULL,
```
Comment (nsnguyen) [inline SQL CODE.txt:35]: video_dislikes can be null too?

File: SQL CODE.txt [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	Video_length TIME NOT NULL,
```
Comment (nsnguyen) [inline SQL CODE.txt:36]: video length can be null too?

File: SQL CODE.txt [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	videoid INT REFERENCES video(videoid) NOT NULL,
+	userid INT REFERENCES users(userid) NOT NULL,
+	tag_name VARCHAR(100)
+)
+
+
+CREATE TABLE video_edits
+(
+	edit_id INT PRIMARY KEY IDENTITY(1,1),
+	Text_Display BIT NOT NULL,
```
Comment (nsnguyen) [inline SQL CODE.txt:69]: why did you want to put these as NOT NULL? I would prefer NULLABLE for these fields too

File: SQL CODE.txt [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+
+
+
+
+CREATE TABLE Categories
+(
+	CategoryID INT PRIMARY KEY IDENTITY(1,1),
+	Product_Review_Videos INT NOT NULL,
```
Comment (nsnguyen) [inline SQL CODE.txt:106]: I think these can be nulls too

**Comment by bensoccer94** [comment]
+
+
+
+
+CREATE TABLE users
+(
+ userId INT PRIMARY KEY IDENTITY(1,1),
+ first_name VARCHAR(40) NOT NULL,
+ last_name VARCHAR(40) NULL,
+ username VARCHAR(25) UNIQUE NOT NULL,
+ email VARCHAR(30) UNIQUE NOT NULL CHECK (email like '%_@%.com'),
+ [password] VARCHAR(15) NOT NULL,






ur right, it makes more sense to use user_password as the column name then
[password]












+
+
+
+CREATE TABLE video
+(
+ videoId INT PRIMARY KEY IDENTITY(1,1),
+ userId INT REFERENCES users(userId) NOT NULL,
+ Video_name VARCHAR(30) NOT NULL,
+ Video_description VARCHAR(MAX) NULL,
+ latitude float NULL,
+ longitude float NULL,
+ video_path VARCHAR(MAX) NOT NULL,
+ Video_Likes INT NOT NULL,





i think the video likes column should be not null so users will know how
many likes they recieved
for each video?









i think the video_dislikes column should be not null for the same reason








+
+
+CREATE TABLE video
+(
+ videoId INT PRIMARY KEY IDENTITY(1,1),
+ userId INT REFERENCES users(userId) NOT NULL,
+ Video_name VARCHAR(30) NOT NULL,
+ Video_description VARCHAR(MAX) NULL,
+ latitude float NULL,
+ longitude float NULL,
+ video_path VARCHAR(MAX) NOT NULL,
+ Video_Likes INT NOT NULL,
+ Video_dislikes INT NOT NULL,
+ Video_length TIME NOT NULL,





the video length column can be null but shouldnt users know how long each
video will be







> +
+
+
+CREATE TABLE tags
+(
+ tag_id INT PRIMARY KEY IDENTITY(1,1),
+ videoid INT REFERENCES video(videoid) NOT NULL,
+ userid INT REFERENCES users(userid) NOT NULL,
+ tag_name VARCHAR(100)
+)
+
+
+CREATE TABLE video_edits
+(
+ edit_id INT PRIMARY KEY IDENTITY(1,1),
+ Text_Display BIT NOT NULL,











i should set the text_display column to null because users wont always use
that feauture
for there videos





the reason why i assigned the user_id and video_id columns to not null
in the tags table is for users to know which type of tags theyve used
for there own videos and which user tagged another user in a video. like
instagram?









> + comment_id INT PRIMARY KEY IDENTITY(1,1),
+ userId INT REFERENCES users(userId) NOT NULL,
+ videoid INT REFERENCES video(videoid),
+ comments VARCHAR(200) NULL,
+
+
+)
+
+
+
+
+
+CREATE TABLE Categories
+(
+ CategoryID INT PRIMARY KEY IDENTITY(1,1),
+ Product_Review_Videos INT NOT NULL,







the reason why the user_id and video_id columns are not null in the
comments table
is so users can know who commented on there video?








product_review_videos column should be not null so will know how many
videos are in that
category?

**Comment by bensoccer94** [comment]
CREATE TABLE video_edits
(
edit_id INT PRIMARY KEY IDENTITY(1,1),
Text_Display BIT NOT NULL,
FastFoward BIT NOT NULL,
Slow_Mo BIT NOT NULL,
Color_Added BIT NOT NULL,

)



CREATE TABLE edited_videos
(
edited_videos_id INT PRIMARY KEY IDENTITY(1,1),
edit_id INT REFERENCES video_edits(edit_id),
videoid INT REFERENCES video(videoid) NOT NULL,
userId INT REFERENCES users(userId) NOT NULL

)



actually, the reason why i used the not null constraint for the
text_display column
is because the edit_id column is a foreign key in the edited_videos table.
This allows the user
to know what edits were made to a video?

**Comment by nsnguyen** [comment]
ah okay. so Video_Likes or Video_dislikes or Slow_Mo are foriegn keys too? The fallback of using too much foreign keys is that if you need to delete one record from a table, it will need to cascade to all other tables that use foreign keys. 

Another fallback of setting NOT NULL to a column is you're putting too much restriction on the column. What happens if the data is really null? does that mean you convert it to 'null' string? It'll be pretty confusing down the road.

For me, i would only set columns to NOT NULL if i know it won't ever be null, like IDs. I think others can be nulls

**Comment by bensoccer94** [comment]
ahhhhh okay okay :)

**Pull request merged**